From sshtylyov@ru.mvista.com Wed May 16 18:11:13 2007
Return-Path: <sshtylyov@ru.mvista.com>
X-Spam-Checker-Version: SpamAssassin 3.1.7-deb (2006-10-05) on debian
X-Spam-Level: 
X-Spam-Status: No, score=0.0 required=5.0 tests=AWL autolearn=unavailable 
	version=3.1.7-deb
Received: from imap.sh.mvista.com (unknown [63.81.120.155]) by mail.tglx.de
	(Postfix) with ESMTP id B7F0D65C065 for <tglx@linutronix.de>; Wed, 16 May
	2007 18:11:13 +0200 (CEST)
Received: from wasted.dev.rtsoft.ru (unknown [10.150.0.9]) by
	imap.sh.mvista.com (Postfix) with ESMTP id 11FC13EC9 for
	<tglx@linutronix.de>; Wed, 16 May 2007 08:38:17 -0700 (PDT)
From: Sergei Shtylyov <sshtylyov@ru.mvista.com>
Organization: MontaVista Software Inc.
To: tglx@linutronix.de
Subject: [PATCH 2.6.21-rt1] ns2cyc() result fix
Date: Wed, 16 May 2007 18:39:50 +0300
User-Agent: KMail/1.5
MIME-Version: 1.0
Content-Disposition: inline
Content-Type: text/plain; charset="iso-8859-1"
Message-Id: <200705161939.50242.sshtylyov@ru.mvista.com>
X-Evolution-Source: imap://tglx%40linutronix.de@localhost:8993/
Content-Transfer-Encoding: 8bit

Fix the dubious use of cycles_t where cycle_t was appropriate.  On the machines
with 32-bit cycles_t (like ARM/PPC) it caused these warnings:

In file included from arch/powerpc/kernel/time.c:1045:
include/linux/clocksource.h: In function `ns2cyc':
include/linux/clocksource.h:213: warning: comparison of distinct pointer types lacks a cast
include/linux/clocksource.h:213: warning: right shift count >= width of type
include/linux/clocksource.h:213: warning: passing argument 1 of `__div64_32' from incompatible pointer type

This function and therefore usecs_to_cycles() was unlikely to return a correct
result on such machines because of the shift result truncation.

Signed-off-by: Sergei Shtylyov <sshtylyov@ru.mvista.com>

---
I'm also uncertain about 'preempt_max_latency' and 'preempt_thresh' variables
being declared as 'unsigned long' -- however, looks like those are unlikely to
overflow... yet it's unclear why there's casts to 'cycle_t' (which is always
64-bit) when initializing/comparing them...

---
 include/linux/clocksource.h |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

Index: linux-2.6.23.1-rt5/include/linux/clocksource.h
===================================================================
--- linux-2.6.23.1-rt5.orig/include/linux/clocksource.h
+++ linux-2.6.23.1-rt5/include/linux/clocksource.h
@@ -189,9 +189,9 @@ static inline s64 cyc2ns(struct clocksou
  * @cs:		Pointer to clocksource
  * @nsecs:	Nanoseconds
  */
-static inline cycles_t ns2cyc(struct clocksource *cs, u64 nsecs)
+static inline cycle_t ns2cyc(struct clocksource *cs, u64 nsecs)
 {
-	cycles_t ret = nsecs << cs->shift;
+	cycle_t ret = nsecs << cs->shift;
 
 	do_div(ret, cs->mult + 1);
 
